File: app/src/main/java/org/kiwix/kiwixmobile/base/BasePresenter.java
/*
 * Kiwix Android
 * Copyright (C) 2018  Kiwix <android.kiwix.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.kiwix.kiwixmobile.base;

import org.kiwix.kiwixmobile.base.contract.Presenter;
import org.kiwix.kiwixmobile.base.contract.ViewCallback;

/**
 * Created by EladKeyshawn on 05/04/2017.
 */

public class BasePresenter<T extends ViewCallback> implements Presenter<T> {

  private T mMvpView;

  @Override
  public void attachView(T mvpView) {
    this.mMvpView = mvpView;
  }

  @Override
  public void detachView() {
    mMvpView = null;
  }

  public boolean isViewAttached() {
    return mMvpView != null;
  }

  public T getMvpView() {
    return mMvpView;
  }


  public void checkViewAttached() {
    if (!isViewAttached()) throw new MvpViewNotAttachedException();
  }

  public static class MvpViewNotAttachedException extends RuntimeException {
    public MvpViewNotAttachedException() {
      super("Please call Presenter.attachView(MvpView) before" +
          " requesting data to the Presenter");
    }
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/base/BasePresenter.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.base

import io.reactivex.disposables.CompositeDisposable
import org.kiwix.kiwixmobile.core.base.BaseContract.Presenter
import org.kiwix.kiwixmobile.core.base.BaseContract.View

/**
 * All presenters should inherit from this presenter.
 */
abstract class BasePresenter<T : View<*>?> : Presenter<T> {
  @JvmField val compositeDisposable = CompositeDisposable()
  @JvmField var view: T? = null

  override fun attachView(view: T) {
    this.view = view
  }

  override fun detachView() {
    view = null
    if (!compositeDisposable.isDisposed) {
      compositeDisposable.dispose()
    }
  }
}
